### PR TITLE
Update proposal for filtering on sub-resources

### PR DESCRIPTION
**Pull request by FelisiaM**
We are planning to use this for service instances

Co-authored-by: Derik Evangelista <devangelista@pivotal.io>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1191,9 +1191,22 @@ GET /apps/:guid?fields=expensive_field
...
+
+This is a mechanism for including multiple fields of related resources in a single response.
+
+Resources and collections **may** accept a `fields[path]` query parameter for each related resources with a list of resource fields. 
+
+Each resource path **MUST** be a series of period-separated relationship names. For example: `app.space.organization`. 
+
+The list of resource fields **MUST** be comma delimited. For example: `fields[space]=name&fields[space.organization]=name,guid`.
+
+Included resource fields **MUST** be returned in an `included` object on the primary resource or collection.
```
Comment (Gerg) [inline README.md:1204]: The example below also shows non-guid fields directly in the relationships, is that also related to `fields`? or will those always be included in the relationship object?

File: README.md [hunk 2]
```diff
@@ -1208,6 +1221,79 @@ GET /v3/apps/:guid?fields=guid,name&fields[droplet]=guid
...
+                        "data": {
+                            "guid": "13c60e38-11e7-11ea-9106-33ee3c5bd4d7"
+                        }
+                    }
+                }
+            }
+        ],
+		"service_offerings": [
+            {
+                "name": "offering-name"
```
Comment (Gerg) [inline README.md:1332]: If there is a list of multiple service offerings, how will clients link these names back to which object they are related to? For included resources, clients can link them together via guids.

File: README.md [hunk 3]
```diff
@@ -1191,9 +1191,22 @@ GET /apps/:guid?fields=expensive_field
 }
 ```
 
-### Proposal: Fields For Sub-Resources
+### Fields For Sub-Resources
```
Comment (Gerg) [inline README.md:1194]: My impression of `fields` (both here and in JSON:API) was to return a subset of already returned data. In that case, you could use `include` + `fields` to return sparse versions of the included resources. 

This proposal (in addition to the "Hidden Fields" proposal above) views `fields` as a way to also include additional data with a request.

File: README.md [hunk 4]
```diff
@@ -1208,6 +1221,80 @@ GET /v3/apps/:guid?fields=guid,name&fields[droplet]=guid
 }
 ```
 
+
+```json
+GET /v3/service_instances?fields[service_plan.service_offering]=name
```
Comment (Gerg) [inline README.md:1225]: It looks like there are also included resources here. Does using `fields` automatically `include` resources for you? Or would you also need to `include` them manually?

File: README.md (hunk 1 above)
Comment (FelisiaM) [reply README.md:1204]: No, that is something that we have been doing in some of the resources for convenience. Can update the example to leave them out so that it is clearer.

File: README.md (hunk 2 above)
Comment (FelisiaM) [reply README.md:1332]: Updated to include the guid as well.

File: README.md (hunk 4 above)
Comment (FelisiaM) [reply README.md:1225]: We tried to follow the JSON API spec on this one and noticed that there is no definitive specification on what happens when `include` is not passed and `fields[related.resource]`is passed.  Given our use-case of only wanting to include certain fields rather than the entire resource we can choose to implement `fields[related.resource]` without the need for `include`. So `fields[related.resource]=name` is saying include the related resource and show only name field.

File: README.md (hunk 3 above)
Comment (FelisiaM) [reply README.md:1194]: What we want to do is something closer to hidden fields, but for related resources. If `fields` query param is used either for space versions or hidden fields as per this proposal, then `fields[related.resource]` can be the equivalent for either sparse fields or hidden fields in a related resource

File: README.md [hunk 5]
```diff
@@ -1191,9 +1191,22 @@ GET /apps/:guid?fields=expensive_field
 }
 ```
 
-### Proposal: Fields For Sub-Resources
+### Fields For Sub-Resources
+
+This is a mechanism for including multiple fields of related resources in a single response.
```
Comment (Gerg) [inline README.md:1196]: The other interesting aspect of this is that the included fields potentially have different permissions than including the whole resource (field-level permissions). So I might be able to  `fields[related.resource]` even if I couldn't `includes=related.resource`. This also means that potentially I could `fields[related.resource]=foo` and not `fields[related.resource]=bar`, even if both of those would work for someone with more permissions (say, a global admin).

File: README.md (hunk 1 above)
Comment (Gerg) [reply README.md:1204]: Is there an advantage to including the name in the relationship vs using the `includes` parameter? Also, if we want to do this more generically, not all resources have `name` fields. Maybe we could use a more generic field like `presentation_name` or something.

Comment (FelisiaM) [reply README.md:1204]: I updated the example to not include the name since it is not something that is done for all resources (we did for some service resources since it's more useful than guid). This PR is not intended to add "name" into the "relationships" part of the resource. Maybe that's something we should discuss separately.

File: README.md (hunk 5 above)
Comment (FelisiaM) [reply README.md:1196]: >The other interesting aspect of this is that the included fields potentially have different permissions than including the whole resource (field-level permissions). So I might be able to fields[related.resource] even if I couldn't includes=related.resource. 

Yes; that's the use-case we are aiming for. For example, domains are shared with organizations, "owners" of domains might have permission to read the org name, but not the rest of the Organization model.

>This also means that potentially I could fields[related.resource]=foo and not fields[related.resource]=bar, even if both of those would work for someone with more permissions (say, a global admin).

Maybe. In our initial implementation, `fields` will be documented and supported in a specific way: all users that can read the resource would be able to use the supported values for `fields`. For example: the only fields one can query on service instances is `fields[service_plan.service_offering]=name,guid`. If a user can list the service instances, they can read the service offering name and guid using (even without access to the service offering resource directly).

**Comment by sweinstein22** [comment]
We are archiving this repo in favor of colocating the information [in the ccng repo](https://github.com/cloudfoundry/cloud_controller_ng/blob/main/docs/v3_style_guide.md), any PRs are welcome in its new location. Thanks!
